web_poet/mixins.py

```diff
@@ -78,3 +81,30 @@ def urljoin(self, url: str) -> str:
         url and baseurl of the response"""
         # FIXME: move it to HttpResponse
         return urljoin(self.base_url, url)
+
+
+class _UrlMixin(abc.ABC):
+    """Common shortcut methods for working with the URL of HTTP requests and
+    responses.
+
+    This mixin could be used with Page Object base classes.
+
+    It requires the ``url`` attribute to be present.
+    """
+
+    _cached_base_url = None
+    url: Union[str, RequestUrl, ResponseUrl]
+
+    @property
+    def _base_url(self) -> str:
+        if self._cached_base_url is None:
+            text = getattr(self, "text", "")[:4096]
+            self._cached_base_url = get_base_url(text, str(self.url))
+        return self._cached_base_url
+
+    def urljoin(self, url: Union[str, RequestUrl, ResponseUrl]) -> RequestUrl:
```

Comment:
Instead of adding it as mixin, I was thinking that `urljoin()`  and `basepath` could perhaps be added in the smart "url" class that we were discussing before.

Reply:
`response.url / arg` syntax? Should `response / arg` also work?

Reply:
@kmike Good point, I haven't considered the `<base>` tag in this case. What do you think about the following behaviors?

1. Using `response.url.join("path/somewhere")` and `response.url / "path/somewhere"` would only take into account the URL value. It's similar to https://yarl.readthedocs.io/en/latest/api.html?highlight=join#yarl.URL.join.

2. When using `response.urljoin("path/somewhere")` _(which has access to the response body/html)_, it adds a new feature of checking the base tag via https://w3lib.readthedocs.io/en/latest/w3lib.html#w3lib.html.get_base_url.

However, there's something that I don't like about the names above, where `response.url.join()` and `response.urljoin()` look very similar and might cause some debugging issues later on.

Reply:
I think regardless of `response.url / "foo"` behavior, it's fine to have response.urljoin, and the existing logic (using base url) is good. So, we can handle ResponseUrl case separately.